pkg/serverless/proxy/transport.go

```diff
@@ -0,0 +1,76 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package proxy
+
+import (
+	"bytes"
+	"errors"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type proxyTransport struct {
+	processor InvocationProcessor
+}
+
+func (p *proxyTransport) RoundTrip(request *http.Request) (*http.Response, error) {
+	log.Debug("[proxy] new request to %s", request.URL)
+
+	// enrich the currentInvocationDetails object
+	enrichCurrentInvocation(p, request)
+
+	response, err := http.DefaultTransport.RoundTrip(request)
+	if err != nil {
+		log.Error("could not forward the request", err)
+		return nil, err
+	}
+
+	dumpedResponse, err := httputil.DumpResponse(response, true)
+	if err != nil {
+		log.Error("could not dump the response")
+		return nil, err
+	}
+
+	// extract only the payload as headers can be retrieved without inspecting the response
+	indexPayload := bytes.Index(dumpedResponse, []byte("\r\n\r\n"))
+	if indexPayload == -1 {
+		return nil, errors.New("invalid payload format")
+	}
+
+	// enrich the currentInvocationDetails when /next response is received
+	if request.Method == "GET" && strings.HasSuffix(request.URL.String(), "/next") {
+		details := &InvocationStartDetails{
+			StartTime:          time.Now(),
+			InvokeHeaders:      response.Header,
+			InvokeEventPayload: string(dumpedResponse[indexPayload:]),
+		}
+		p.processor.OnInvokeStart(details)
+	}
+
+	return response, nil
+}
+
+func enrichCurrentInvocation(p *proxyTransport, request *http.Request) {
```

Comment:
I think you might want to rename this function given the new approach. It's not really enriching anything any more, it seems it just calls the appropriate hook.